### PR TITLE
feat(datasource/conda): add conda datasource

### PR DESCRIPTION
**Pull request by morremeyer**
## Changes

This PR adds a datasource for conda as discussed in #2213. 

## Context

Issue #2213. This PR starts with the datasource, a manager can be added later on.

## Documentation (please check one with an [x])

- [x] I have updated the documentation, or
- [ ] No documentation update is required

## How I've tested my work (please tick one)

I have verified these changes via:

- [ ] Code inspection only, or
- [x] Newly added/modified unit tests, or
- [ ] No unit tests but ran on a real repository, or
- [ ] Both unit tests + ran on a real repository

### CODE REVIEW

**Comment by morremeyer** [comment]
I've drafted this PR for now as it will fail tests and the tests will likely need to be updated, too.

If someone could lend me a pair of eyes on the issue here, I think I’m missing something obvious. To reproduce locally:

```sh
git clone git@github.com:anaconda-hackdays/renovate.git
git checkout feat/2213-datasource-conda
yarn install --frozen-lockfile
yarn jest lib/datasource/conda
```

I get an error for the mocks

```txt
    Pending mocks!
     * GET [https://api.anaconda.org:443/package/main/pytest](https://api.anaconda.org/package/main/pytest)
```

on all the tests. Afaik, this means that the mock is not used. But I fail to see why it is not used.

File: lib/datasource/conda/__snapshots__/index.spec.ts.snap [hunk 1]
```diff
@@ -0,0 +1,3 @@
+// Jest Snapshot v1, https://goo.gl/fbAQLP
+
+exports[`datasource/conda/index getReleases processes real data 1`] = `null`;
```
[the remark this answers is not in the document]
Comment (morremeyer) [reply lib/datasource/conda/__snapshots__/index.spec.ts.snap:3]: I plan to update this one, however right now this is how jest generates it. I have a strong suspicion that this has to do with my comment above (https://github.com/renovatebot/renovate/pull/14257#issuecomment-1041580178) as this test also has the `Pending mocks!` issue.

[the remark this answers is not in the document]
Comment (viceice) [reply lib/datasource/conda/__snapshots__/index.spec.ts.snap:3]: So your code didn't work as expected. 😉

File: lib/datasource/conda/index.ts [hunk 2]
```diff
@@ -0,0 +1,65 @@
...
+
+    const url = joinUrlParts(registryUrl, lookupName);
+
+    const result: ReleaseResult = {
+      releases: [],
+    };
+    try {
+      const response = await this.http.get(url);
+
+      const packageJson = JSON.parse(response.body);
```
Comment (viceice) [inline lib/datasource/conda/index.ts:42]: Do not manually parse JSON, use `http.getJson` function and create a proper type.

File: lib/datasource/conda/index.spec.ts [hunk 3]
```diff
@@ -0,0 +1,65 @@
...
+        getPkgReleases({
+          datasource,
+          depName,
+        })
+      ).rejects.toThrow(EXTERNAL_HOST_ERROR);
+    });
+
+    it('processes real data', async () => {
+      httpMock
+        .scope(defaultRegistryUrl)
```
Comment (viceice) [inline lib/datasource/conda/index.spec.ts:54]: The url needs ending slash to get proper scope. 😉

Comment (morremeyer) [reply lib/datasource/conda/index.spec.ts:54]: Why?

nock itself doesn't need that as per its documentation (https://github.com/nock/nock#usage) and I don't see [the `scope` code](https://github.com/renovatebot/renovate/blob/b6c374f16698b1a20c01cdeb1de353634a698dc9/test/http-mock.ts#L60-L82) treating it differently.

However, it looks to me like nock strips out the trailing slash anyways, as the following spec **with** a trailing slash at the `baseUrl` still is missing a slash.

```ts
    it('test with literals', async () => {
      httpMock
        .scope('https://api.anaconda.org/package/')
        .get('main/pytest')
        .reply(200, Fixtures.get('pytest.json'));
      const res = await getPkgReleases({
        datasource,
        depName,
      });
      expect(res).toMatchSnapshot();
      expect(res?.releases).toHaveLength(94);
    });
```

Will result in

```
  ● datasource/conda/index › getReleases › test with literals

    Pending mocks!
     * GET https://api.anaconda.org:443/packagemain/pytest
```

File: lib/datasource/conda/__snapshots__/index.spec.ts.snap (hunk 1 above)
[the remark this answers is not in the document]
Comment (morremeyer) [reply lib/datasource/conda/__snapshots__/index.spec.ts.snap]: After staring at things and attaching a debugger (duh, could've done that way earlier 😆 ) I found the issue, fixed in 688687efa5d755d470ac55e837c7ed7099a9a678.

I'll work on the test coverage now.

File: lib/datasource/conda/index.spec.ts (hunk 3 above)
Comment (morremeyer) [reply lib/datasource/conda/index.spec.ts:54]: I'll resolve this as I updated the config and it makes sense to do so, but if you can provide more insight, that would be great @viceice.

**Comment by morremeyer** [comment]
I solved the `Pending mocks!` issue, so this is now ready for review.

**Comment by viceice** [comment]
> I solved the `Pending mocks!` issue, so this is now ready for review.

https://github.com/renovatebot/renovate/blob/main/.github/contributing.md#re-requesting-a-review 😉

File: lib/datasource/conda/__fixtures__/readme.md [hunk 4]
```diff
@@ -0,0 +1,3 @@
+# conda fixtures
```
Comment (HonkingGoose) [inline lib/datasource/conda/__fixtures__/readme.md:1]:
```suggestion
# Conda fixtures
```

File: lib/datasource/conda/__fixtures__/readme.md [hunk 5]
```diff
@@ -0,0 +1,3 @@
...
+
+`pytest.json` contains a response body from https://api.anaconda.org/package/main/pytest, but the `files` key has been truncated to one element as it is huge and not needed for renovate.
```
Comment (HonkingGoose) [inline lib/datasource/conda/__fixtures__/readme.md:3]:
```suggestion
`pytest.json` contains a response body from https://api.anaconda.org/package/main/pytest, but the `files` key has been truncated to one element as it is huge and not needed for Renovate.
```

File: lib/datasource/conda/readme.md [hunk 6]
```diff
@@ -0,0 +1,7 @@
+# conda
```
Comment (HonkingGoose) [inline lib/datasource/conda/readme.md:1]:
```suggestion
# Conda
```

File: lib/datasource/conda/readme.md [hunk 7]
```diff
@@ -0,0 +1,7 @@
...
+
+This datasource returns releases from the specified [conda](https://docs.conda.io/en/latest/) registry.
```
Comment (HonkingGoose) [inline lib/datasource/conda/readme.md:3]:
```suggestion
This datasource returns releases from the specified [Conda](https://docs.conda.io/en/latest/) registry.
```

File: lib/datasource/conda/types.ts [hunk 8]
```diff
@@ -0,0 +1,91 @@
+/* For values of unknown type that are not used with renovate,
+   any object is accepted in the types defined here
+*/
+
+// This can be improved with the documentation at
+// https://docs.conda.io/projects/conda/en/latest/user-guide/concepts/pkg-specs.html#supported-version-strings
+export type CondaVersion = string;
```
Comment (viceice) [inline lib/datasource/conda/types.ts:7]: Not sure if we really need this, but ok to have. 🤗

File: lib/datasource/conda/readme.md [hunk 9]
```diff
@@ -0,0 +1,7 @@
...
+
+The default registry is `https://api.anaconda.org/package/`, which queries a specific conda channel for a specific package.
```
Comment (HonkingGoose) [inline lib/datasource/conda/readme.md:5]:
```suggestion
The default registry is `https://api.anaconda.org/package/`, which queries a specific Conda channel for a specific package.
```

**Review by HonkingGoose — CHANGES_REQUESTED** [review]

**Review by viceice — DISMISSED** [review]

**Comment by morremeyer** [comment]
@HonkingGoose I asked some of the conda contributors with Anaconda, feedback is „It's conda, but capitalized at the start of sentences.“

This can be seen e.g. in https://docs.conda.io/projects/conda/en/latest/glossary.html#conda

**Review by viceice — APPROVED** [review]

**Comment by viceice** [comment]
> Yes, anaconda-hackdays is one of our organizations. Should I reopen the PR from my personal fork?

No, but future PR's, otherwise it can defer merges if we maintainer can't update the branch. It must be up-to-date to merge. 😉

**Comment by morremeyer** [comment]
That makes sense, thanks for letting me know. Should I update the branch with „Update branch“ in the meantime? 

Also, can I do anything about the Code scanning pipeline failure or is that not needed?

**Comment by viceice** [comment]
Yes, update branch will retrigger all checks and should solve the code scan issue

**Comment by rarkins** [comment]
Thanks, @morremeyer

**Comment by astellingwerf** [comment]
Could the H1 at the first line of the readme.md be the reason that the menu in the documentation stops after it?:
image

**Pull request merged**